Merge 2021 Infinite Recharge @ Home changes

FireShredder24 opened the pull request:
Refactors or overhauls:

- MoPrefs
- Shooter verification
- Shuffleboard widgets
- drive PID
- Shooter safety

Adds:

- Drivetrain simulation
- Path following
- Automatic shooting w/ Limelight

Tweaks:

- Drive handling (deadzones, PID)
- Shooter PID
- Controller bindings (Xbox + F310 -> F310 alone)

FireShredder24 committed:
Replace SmartDashboard widgets

Replace existing SmartDashboard status widgets with the appropriate Shuffleboard widgets

FireShredder24 committed:
Add PID Widgets for Shooter

FireShredder24 committed:
Add PID Widgets for Drive

Add constant updater method to periodic() of both FalconDriveSubsystem and ShooterSubsystem

FireShredder24 committed:
Remove duplicates, change speed widget to graph

FireShredder24 committed:
Initiate PathWeaver project

FireShredder24 committed:
Build test path

FireShredder24 committed:
Add blank PathWeaverCommand

Refactor AutoChooser to take any number of Commands

FireShredder24 committed:
Add blank PathWeaverCommand, for real

FireShredder24 committed:
Fetch trajectory paths

FireShredder24 committed:
Add blank override methods

terbos90803 committed:
Meeting changes

FireShredder24 committed:
Add overload of drive() for RAMSETE

FireShredder24 committed:
Remove unnecessary method

Also add comments to drive(ChassisSpeeds)

FireShredder24 committed:
Add comments

FireShredder24 committed:
Remove unnecessary comments

FireShredder24 committed:
Revert hood verification logic

Also, remove unnecessary setHoodPos(double)

FireShredder24 committed:
Remove unnecessary PID settings from shooter

Add more shooter hood comments

jordan-powers committed:
Fix graphs on SmartDashboard

Turns out that the SmartDashboard isn't fond of special characters in the
NetworkTables keys.

jordan-powers committed:
Remove turn request negation from DriveCommand

A negation was added in DriveCommand as a quick fix for an inverted turn
request. The proper way to fix this is to use a DriveConditioner.
This fix reverts the negation in DriveCommand and adds a new
DriveConditioner that inverts the turn request.

This fixes the broken DriveCommandTest unit test.

jordan-powers committed:
ShooterHoodSubsystem: Update hoodPos in periodic()

FireShredder24 committed:
Revert shooter setpoint to a MoPref

Add comment in ControllerBase

Update MAX_FREE_SPEED to match observed maximum free speed

FireShredder24 committed:
Replace shooter PID constants with MoPrefs

FireShredder24 committed:
Remove empty NetworkTableEntries

FireShredder24 committed:
Replace Drive PID widgets with MoPrefs

FireShredder24 committed:
Replace ShooterHood PID Constants with MoPrefs

FireShredder24 committed:
Fix new code smells

jordan-powers committed:
Add MoPrefsKey enum

MoPrefs are now accessed using the get() method, which takes a MoPrefsKey.
MoPrefsKeys are responsible for tracking the default value of their
associated preference, and are used as keys into the Preferences table.

This improves maintainability: adding a new preference only requires
adding a new entry to the MoPrefsKey enum instead of a whole new method
on MoPrefs.

jordan-powers committed:
Finish conversion to using MoPrefsKey enum

jordan-powers committed:
Merge pull request #27 from momentumfrc/moprefs-enum

Add MoPrefsKey enum

FireShredder24 committed:
Add logic for calculating robot pose

Also, add logic for passing Trajectory.State to a RamseteController

FireShredder24 committed:
Fix pose generation

Implement trajectory tracking logic in PathWeaverCommand

Add unit tests for drive calculations

FireShredder24 committed:
Merge remote-tracking branch 'origin/shuffleboard' into path-weaver

jordan-powers committed:
Show full test failure output

FireShredder24 committed:
Fix failing tests

Removes the call of updatePidConstants() from the constructors of all subsystems that use it.  It was being called before robotInit() during deploy.

FireShredder24 committed:
Add characterization data

FireShredder24 committed:
Fix tests and fix generatePose()

jordan-powers commented on characterization/characterization-project/build/tmp/jar/MANIFEST.MF:
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
```
Add `build/` to the .gitignore
Use `git rm` to remove these temporary build files.
In fact, if this is just a separate FRC java project, you could just probably just copy the .gitignore from the main code to this project's directory.

jordan-powers replied:
Nevermind, I got sick of scrolling past these files so I did it

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -95,29 +98,32 @@
-      new ParallelCommandGroup(
-          new RunCommand(() -> shooterSubsystem.shoot(MoPrefs.getShootFromWallHoodSetpoint()), shooterSubsystem)),
+      new ParallelCommandGroup(new RunCommand(
+          () -> shooterSubsystem.shoot(MoPrefs.get(MoPrefsKey.SHOOT_FROM_WALL_HOOD_SETPOINT)), shooterSubsystem)),
```
This looks like it was branched off of #26.
If that's the case, then we'll probably want to avoid merging it until #26 is closed, since this is going to have all the same changes.

jordan-powers commented on src/main/java/frc/robot/commands/PathWeaverCommand.java:
```diff
@@ -0,0 +1,60 @@
+import frc.robot.subsystems.FalconDriveSubsystem;
+
+public class PathWeaverCommand extends CommandBase {
+  private static final String TRAJECTORY_PATH = "paths/test.path";
```
Hard coded path?
We'll probably want to use a SendableChooser for this at some point

jordan-powers commented on src/main/java/frc/robot/commands/PathWeaverCommand.java:
```diff
@@ -0,0 +1,60 @@
+      startTime = Timer.getFPGATimestamp();
+    }
+    // Calculate the time since the start of trajectory tracking
+    double elapsedTime = Timer.getFPGATimestamp() - startTime;
```
Why didn't you use a Timer object for this?

jordan-powers commented on src/main/java/frc/robot/commands/PathWeaverCommand.java:
```diff
@@ -0,0 +1,60 @@
+  private final FalconDriveSubsystem subsystem;
+  private final RamseteController controller = new RamseteController();
+
+  public PathWeaverCommand(FalconDriveSubsystem subsystem) {
```
Might be a good idea to take the Path as a constructor argument. Makes the code usable with multiple different paths, including potentially paths calculated on-the-fly

jordan-powers commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -185,4 +169,75 @@ public void stop() {
+  }
+
+  public Pose2d getPose() {
+    return generatePose(getWheelVelocity(leftFront.getSensorCollection().getIntegratedSensorVelocity()),
```
You're calling getWheelVelocity twice: here and on line 216.

jordan-powers commented on src/main/java/frc/robot/subsystems/FalconDriveSubsystem.java:
```diff
@@ -185,4 +169,75 @@ public void stop() {
+  public double getWheelVelocity(double encoderVelocity) {
+    double encTicksPerSecond = encoderVelocity * 10; // Talon FX outputs in encoder ticks per 100ms, but we want it per
+                                                     // second
+    double rotationsPerSecond = encTicksPerSecond / 2048; // 2048 encoder ticks = 1 rotation
```
Magic number

jordan-powers commented on src/test/java/frc/robot/subsystems/FalconDriveSubsystemTest.java:
```diff
@@ -0,0 +1,46 @@
+  public void testPoseGen() {
+    // Represents the robot turning in place
+    Pose2d pose = m_subsystem.generatePose(2048, -2048);
+    System.out.println("Rot: " + pose.getRotation().getRadians());
```
I'm pretty sure it's best practice to only have tests output information if the test fails.

jordan-powers committed:
Fix .gitignore

FireShredder24 committed:
Add SendableChooser for paths

Remove unnecessary calls to getWheelVelocity()

Remove value printing on successful test

Add path updater method

FireShredder24 committed:
Meeting changes

FireShredder24 committed:
Cleanup meeting changes

Remove debugging prints

Add comments

FireShredder24 committed:
Address code smells

FireShredder24 committed:
Add Barrel Racing path

FireShredder24 committed:
Add Barrel Racing path

FireShredder24 committed:
Update to WPILib 2021.2.2

FireShredder24 committed:
Fix Field2d

FireShredder24 committed:
Simulation refactor WIP

Refactors FalconDriveSubsystem so that the Talon FX motor controllers can be simulated.

jordan-powers committed:
Pose2D takes in distances, not velocities

FireShredder24 committed:
Fix loop overruns

Moves the path updater to DisabledInit() so that it is not run periodically, which is very unnecessary
Adds simulation checks to the PID constant updaters of ShooterSubsystem and ShooterHoodSubsystem.

terbos90803 committed:
Merge stashed mtg changes from 2/17

FireShredder24 committed:
Meeting changes

jordan-powers committed:
Custom MoPrefs

This patch removes MoPrefs's use of wpilib's preferences and instead
directly interacts with the network tables. This allows us to access the
underlying NetworkTableEntries, so that we can add update listeners.

jordan-powers committed:
Only update PID constants when they change

FireShredder24 committed:
Address code smells

FireShredder24 committed:
Add NavX support

FireShredder24 committed:
Clean unused isReal checks

Suppress unused warning on encTicksToMeters(double)

FireShredder24 committed:
Fix Limelight auto-aiming

FireShredder24 committed:
Add shooting logic to AutonDriveCommand

FireShredder24 committed:
Fix sim friction and gyro

Reset simulated position upon disable

FireShredder24 committed:
Fix sim auto behavior

Sim auto now behaves similarly to real life

terbos90803 committed:
Meeting changes

FireShredder24 committed:
wip

terbos90803 committed:
Meeting changes

jordan-powers committed:
Fix drive velocity pid

FireShredder24 committed:
Merge John's local changes

FireShredder24 committed:
Add real robot simulation constants

FireShredder24 committed:
Fix sim auto starting position

Cleanup getAutonomousCommand

Update Barrel Racing path

terbos90803 committed:
Fix real-robot polarity

FireShredder24 committed:
Add other PathWeaver paths

terbos90803 committed:
Meeting changes (IR@Home Final!)

FireShredder24 committed:
Merge local changes into upstream

FireShredder24 committed:
Prevent shooter firing with raised intake

terbos90803 committed:
Fix intake safety

FireShredder24 committed:
Set PID constants for shooter and hood

FireShredder24 committed:
Update PID constants

terbos90803 committed:
Cleanup PID prefs

Put defaults in MoPrefs.
Add/Cleanup a lot of comments.

FireShredder24 committed:
Add shooter hood range function

FireShredder24 committed:
Meeting changes

FireShredder24 committed:
Meeting changes

FireShredder24 committed:
Fix Limelight shooting

FireShredder24 committed:
Stow shooter hood if unreliable zero when firing

terbos90803 commented:
This branch is well past its intended purpose and should be merged asap.
Unless there are any protests, I'd like to do that before next Wednesday's meeting.

FireShredder24 committed:
Refactor Shuffleboard and arrange widgets

FireShredder24 committed:
Remove unused imports

terbos90803 reviewed (approved):
Long overdue merge.
